### PR TITLE
Add es prometheus exporter

### PR DESCRIPTION
**Pull request by chenchik**
#### What this PR does / why we need it:

This PR creates a sidecar Prometheus exporter container to the Elasticsearch client service. It also adds a servicemonitor that can be enabled in order to communicate with Prometheus. 

This PR is very useful because it makes it easy to connect Elasticsearch to Prometheus. Prometheus can be used to consolidate logging of all services and applications, including ones outside of the elastic stack. 

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Special notes for your reviewer:

This is more of an additional feature than a bug fix. I think it's useful to provide people to an array of monitoring options. This PR adds an additional monitoring option for people who want to monitor their deployment of the elasticsearch helm chart.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by davidkarlsen** [comment]
Just making you aware that there exists a chart for this in: https://github.com/helm/charts/tree/master/stable/elasticsearch-exporter already

**Comment by chenchik** [comment]
This PR would bundle an exporter and servicemonitor with elasticsearch within the same chart, making it super convenient to work with prometheus.

**Comment by rendhalver** [comment]
I tend to agree with @davidkarlsen. If there is already a chart for it we should just use that.
`stable/elasticserarch-exporter` does look like a good candidate to add as a requirement to the `incubator/elastic-stack` chart.

**Comment by davidkarlsen** [comment]
@rendhalver +1 to that - adding it as an opt-in in the stack which is a complete chart for it all

**Comment by rendhalver** [comment]
I will run up a PR for that.
@chenchik Are you using the `elastic-stack` chart?

**Comment by chenchik** [comment]
I'm using the elasticsearch chart specifically. Outside of the exporter, the servicemonitor object is the other addition to this chart with the PR. Could we incorporate that part of the PR?

**Comment by rendhalver** [comment]
Do you use Kibana and logstash on top of Elasticsearch?

File: incubator/elasticsearch/templates/client-deployment.yaml [hunk 1]
```diff
@@ -131,6 +131,17 @@ spec:
           subPath: elasticsearch.keystore
           readOnly: true
 {{- end }}
+      - name: {{ .Values.prometheusExporter.clientName }}
```
Comment (stealthybox) [inline incubator/elasticsearch/templates/client-deployment.yaml:135]: The metrics sidecar should be optional.
I suggest adding `Values.prometheusExporter.enabled` and using that to wrap the container in an if.

It should probably default to false.

File: incubator/elasticsearch/templates/client-svc.yaml [hunk 2]
```diff
@@ -18,6 +18,10 @@ spec:
     - name: http
       port: 9200
       targetPort: http
+    - name: {{ .Values.prometheusExporter.clientName }}
```
Comment (stealthybox) [inline incubator/elasticsearch/templates/client-svc.yaml:22]: Same here: the service port can be conditionally declared with `Values.prometheusExporter.enabled`

File: incubator/elasticsearch/Chart.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 name: elasticsearch
 home: https://www.elastic.co/products/elasticsearch
-version: 1.8.2
+version: 1.8.3
```
Comment (stealthybox) [inline incubator/elasticsearch/Chart.yaml]: This should be a minor version bump instead of a patch.

File: incubator/elasticsearch/values.yaml [hunk 4]
```diff
@@ -120,5 +120,25 @@ data:
...
+  # The values within here are responsible for setting up service ports and sidecar exporter containers.
+  repository: justwatch/elasticsearch_exporter
+  tag: 1.0.2
+  pullPolicy: "IfNotPresent"
+  clientName: client-exporter
+  internalPort: 9108
+  externalPort: 9108
+  metricsEndpoint: /metrics
+
+serviceMonitor:
```
Comment (stealthybox) [inline incubator/elasticsearch/values.yaml:138]: This option could benefit from a comment linking it to prometheus-operator / kube-prometheus

**Comment by stealthybox** [comment]
@rendhalver For the use-case that inspired this monitoring work, ElasticSearch is used standalone as a component of an application without Logstash/Fluentd.
_________________________
A difference between this patch and the standalone exporter chart is that the exporter is a sidecar on the client node.
This reduces operator config as you don't have to synchronize your Service names for the exporter chart installation.

It's also colocated with the client api pod, but that's not a huge differentiator for elasticsearch as it could be for other apps since the client is already external to the data source.
(This makes a big difference though if you use this chart to run as a single node cluster, though.)
_________________________
The ServiceMonitor addition is a notable opt-in feature of this patch as it makes integration with kube-prometheus much easier. This is inspired by prior art from [stable/rabbitmq-ha](https://github.com/helm/charts/tree/master/stable/rabbitmq-ha#configuration).
When we were consuming the HA RabbitMQ chart, these options made it trivial to integrate with our monitoring stack.

I don't see a major issue with adding the exporter and servicemonitor to this chart as long as both features are optional, as they should be high value and low maintenance.

**Comment by chenchik** [comment]
No, I'm using Elasticsearch as part of an API component. I need to collect metrics related to application health/performance but I don't need to collect logs with Kibana (that is already taken care of by something else).

**Comment by desaintmartin** [comment]
Bundling this as a sidecar is what is already done in postgres, which is quite handy, efficient and simple. I would be in favor of not extending the existing exporter chart.

About the ServiceMonitor, it would allow people like me to not having to extend elasticsearch chart just to add it.

**Comment by desaintmartin** [comment]
Should we close this in favor of using elastic-stack?

**Comment by rendhalver** [comment]
Yes. I opened #8616 on elastic-stack to add the exporter.

**Comment by desaintmartin** [comment]
Note that since our last discussion, redis and rabbitmq charts adopted the side-car container approach about exporter.

**Comment by rendhalver** [comment]
Sounds like a good plan to me @desaintmartin.
I will try and find some time to replicate that into the exporter chart.

**Comment by stealthybox** [comment]
Other charts have set a precedence of using the sidecar.
IMO it's useful to have this option, and the maintenance burden is minimal.

**Pull request closed without merging**